### ISSUE TITLE
clippy::nonminimal_bool suggests a semantics-altering transformation

### ISSUE THREAD

**Issue by RalfJung** [rust-lang/rust-clippy#12371]
### Description

This program does not panic:
```rust
fn test(x: usize) -> bool {
    !x != 0
}

fn main() {
    assert!(test(13));
}
```
But if I run nightly clippy it says
```
warning: this boolean expression can be simplified
 --> src/main.rs:2:5
  |
2 |     !x != 0
  |     ^^^^^^^ help: try: `x == 0`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#nonminimal_bool
  = note: `#[warn(clippy::nonminimal_bool)]` on by default
```
Upon changing the code as requested, it [panics](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=187ac4ca0da69875c0211960e0c6f5d9).

Note that `!x` is *bitwise* negation here, so `(!x) != 0` and `x == 0` are very different conditions.

### Version

```text
rustc 1.78.0-nightly (ef324565d 2024-02-27)
binary: rustc
commit-hash: ef324565d071c6d7e2477a195648549e33d6a465
commit-date: 2024-02-27
host: x86_64-unknown-linux-gnu
release: 1.78.0-nightly
LLVM version: 18.1.0
```

**Comment by GuillaumeGomez** [comment]
It should have checked that all members are booleans before suggesting this. I'll send a fix.
